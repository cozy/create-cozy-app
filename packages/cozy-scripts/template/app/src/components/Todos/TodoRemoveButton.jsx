import React, { Component } from 'react'

import Button from 'cozy-ui/react/Button'
import Icon from 'cozy-ui/react/Icon'
import { withClient } from 'cozy-client'

export class TodoRemoveButton extends Component {
  constructor(props) {
    super(props)
    // initial component state
    this.state = { isWorking: false }
  }

  // delete the related todo
  removeTodo = async () => {
    const { client, todo } = this.props
    this.setState({ isWorking: true })

    try {
      await client.destroy(todo)
    } finally {
      this.setState({ isWorking: false })
    }
  }

  render() {
    const { isWorking } = this.state
    return (
      <Button
        className="todo-remove-button"
        theme="danger"
        iconOnly
        label="Delete"
        busy={isWorking}
        disabled={isWorking}
        onClick={this.removeTodo}
        extension="narrow"
      >
        {!isWorking ? <Icon icon="trash" /> : null}
      </Button>
    )
  }
}

// get mutations from the client to use deleteDocument
export default withClient(TodoRemoveButton)
