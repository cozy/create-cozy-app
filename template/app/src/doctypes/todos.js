export const TODOS_DOCTYPE = 'io.mocks.todos'
