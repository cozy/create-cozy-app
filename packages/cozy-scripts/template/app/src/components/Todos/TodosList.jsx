import React from 'react'

import TodoRemoveButton from './TodoRemoveButton'

export const TodosList = props => {
  const { todos } = props
  if (!todos || !todos.length) return null
  return (
    <div>
      <h2>Todos list</h2>
      <ul className="todos-list">
        {todos.map(todo => (
          <li key={todo._id} className="todo-item">
            <span>{todo.name}</span>
            <TodoRemoveButton todo={todo} />
          </li>
        ))}
      </ul>
    </div>
  )
}

export default TodosList
