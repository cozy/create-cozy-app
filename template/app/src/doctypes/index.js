import { TODOS_DOCTYPE } from './todos'

// the documents schema, necessary for CozyClient
export default {
  todos: {
    doctype: TODOS_DOCTYPE,
    attributes: {},
    relationships: {}
  }
}

// export all doctypes for the application
export * from './todos'
