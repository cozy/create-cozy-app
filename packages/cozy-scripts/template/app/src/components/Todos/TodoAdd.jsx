import React, { Component } from 'react'

import { withClient } from 'cozy-client'
import Input from 'cozy-ui/react/Input'
import Label from 'cozy-ui/react/Label'
import Button from 'cozy-ui/react/Button'

import { TODOS_DOCTYPE } from 'doctypes'

export class TodoAdd extends Component {
  constructor(props, context) {
    super(props, context)
    // initial component state
    this.state = {
      todoToAdd: '',
      isWorking: false
    }
  }

  // handle input value change
  handleChange = event => {
    this.setState({ todoToAdd: event.target.value })
  }

  // create the new todo
  handleSubmit = async () => {
    const { todoToAdd } = this.state
    const { client } = this.props
    // reset the input and display a spinner during the process
    this.setState({ todoToAdd: '' })

    try {
      this.setState({ isWorking: true })
      const newDoc = { _type: TODOS_DOCTYPE, name: todoToAdd }
      await client.save(newDoc)
    } finally {
      this.setState(() => ({ isWorking: false }))
    }
  }

  render() {
    const { todoToAdd, isWorking } = this.state
    return (
      <div>
        <h2 className="u-mt-2 u-mb-half">Add a new Todo</h2>
        <form onSubmit={this.handleSubmit}>
          <Label htmlFor="todo-add-input">Todo name</Label>
          <Input
            value={todoToAdd}
            onChange={this.handleChange}
            id="todo-add-input"
          />
          <Button
            onClick={this.submit}
            type="submit"
            busy={isWorking}
            label="add"
            size="large"
            extension="narrow"
          />
        </form>
      </div>
    )
  }
}

// get mutations from the client to use createDocument
export default withClient(TodoAdd)
