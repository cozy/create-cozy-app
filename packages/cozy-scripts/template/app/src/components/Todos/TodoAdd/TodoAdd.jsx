import React, { useState } from 'react'

import { useClient } from 'cozy-client'
import Input from 'cozy-ui/transpiled/react/Input'
import Label from 'cozy-ui/transpiled/react/Label'
import Button from 'cozy-ui/transpiled/react/Button'

import { TODOS_DOCTYPE } from '../../../doctypes'

export const TodoAdd = () => {
  const client = useClient()
  const [todoToAdd, setTodoToAdd] = useState('')
  const [isWorking, setIsWorking] = useState(false)

  const handleChange = event => {
    setTodoToAdd(event.target.value)
  }

  const handleSubmit = async () => {
    // reset the input and display a spinner during the process
    setIsWorking(true)
    setTodoToAdd('')
    await client.save({ _type: TODOS_DOCTYPE, name: todoToAdd })
    setIsWorking(false)
  }

  return (
    <div>
      <h2>Add a new Todo:</h2>
      <form onSubmit={handleSubmit}>
        <Label htmlFor="todo-add-input"> Todo name: </Label>
        <Input value={todoToAdd} onChange={handleChange} id="todo-add-input" />
        <Button
          type="submit"
          busy={isWorking}
          label="add"
          size="large"
          extension="narrow"
        />
      </form>
    </div>
  )
}

export default TodoAdd
