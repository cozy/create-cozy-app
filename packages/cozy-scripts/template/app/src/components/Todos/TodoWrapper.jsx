import React from 'react'

import { useQuery, isQueryLoading } from 'cozy-client'
import Spinner from 'cozy-ui/transpiled/react/Spinner'

import { getAllTodos } from 'src/utils/queries'
import TodoAdd from 'src/components/Todos/TodoAdd/TodoAdd'
import TodoList from 'src/components/Todos/TodoList/TodoList'

export const TodoWrapper = () => {
  const { data, ...rest } = useQuery(
    getAllTodos.definition,
    getAllTodos.options
  )

  return isQueryLoading(rest) ? (
    <Spinner size="xxlarge" className="u-flex u-flex-justify-center" />
  ) : (
    <>
      <TodoList todos={data} />
      <TodoAdd />
    </>
  )
}

export default TodoWrapper
