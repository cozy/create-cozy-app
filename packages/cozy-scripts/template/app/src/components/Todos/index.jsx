import React from 'react'

import { useQuery, isQueryLoading } from 'cozy-client'
import Spinner from 'cozy-ui/transpiled/react/Spinner'

import { getAllTodos } from '../../utils/queries'
import TodoAdd from './TodoAdd/TodoAdd'
import TodosList from './TodoList/TodosList'

export const Todos = () => {
  const { data, ...rest } = useQuery(
    getAllTodos.definition,
    getAllTodos.options
  )

  return (
    <div className="todos">
      {isQueryLoading(rest) ? (
        <Spinner size="xxlarge" middle />
      ) : (
        <div>
          <TodosList todos={data} />
          <TodoAdd />
        </div>
      )}
    </div>
  )
}

export default Todos
