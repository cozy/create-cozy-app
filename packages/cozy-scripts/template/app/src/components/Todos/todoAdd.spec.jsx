'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

// we import the not connected component version for testing
import { TodoAdd } from 'components/Todos/TodoAdd'
import { TODOS_DOCTYPE } from 'doctypes'
import Input from 'cozy-ui/react/Input'

// mock data for testing
import mockTodos from './_mocksTodos.json'

const created = mockTodos[0]
// async mock function
const mockCreateTodo = jest.fn(() => Promise.resolve(created))

describe('TodoAdd component:', () => {
  beforeEach(() => {
    // reset all jest mock calls data before each test
    jest.resetAllMocks()
  })

  const setup = () => {
    const client = {
      save: mockCreateTodo
    }
    const component = shallow(<TodoAdd client={client} />)
    return { component }
  }

  it('should be rendered correctly', () => {
    const { component } = setup()
    expect(component.getElement()).toMatchSnapshot()
  })

  it('should handle isWorking correctly (display spinner)', () => {
    const { component } = setup()

    expect(component.state().isWorking).toBe(false)
    component.setState({ isWorking: true })
    expect(component.getElement()).toMatchSnapshot()
  })

  it('should handle createDocument correctly on button click', async () => {
    const { component } = setup()
    // set a todo name value
    component.setState({ todoToAdd: created.name })

    const addButton = component.find('form')
    await addButton.simulate('submit')

    expect(mockCreateTodo.mock.calls.length).toBe(1)
    expect(mockCreateTodo.mock.calls[0][0]).toEqual({
      _type: TODOS_DOCTYPE,
      name: 'Todo 1'
    })
  })

  it('should update the state on input value change', () => {
    const { component } = setup()
    const input = component.find(Input)

    expect(component.state().todoToAdd).toBe('')
    input.simulate('change', { target: { value: created.name } })
    expect(component.state().todoToAdd).toBe(created.name)
  })
})
