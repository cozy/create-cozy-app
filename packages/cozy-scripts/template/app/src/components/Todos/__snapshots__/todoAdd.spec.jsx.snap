// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdd component: should be rendered correctly 1`] = `
<div>
  <h2
    className="u-mt-2 u-mb-half"
  >
    Add a new Todo
  </h2>
  <form
    onSubmit={[Function]}
  >
    <Label
      block={true}
      className=""
      error={false}
      htmlFor="todo-add-input"
    >
      Todo name
    </Label>
    <Input
      error={false}
      fullwidth={false}
      id="todo-add-input"
      onChange={[Function]}
      size="large"
      type="text"
      value=""
    />
    <DefaultButton
      busy={false}
      extension="narrow"
      label="add"
      size="large"
      type="submit"
    />
  </form>
</div>
`;

exports[`TodoAdd component: should handle isWorking correctly (display spinner) 1`] = `
<div>
  <h2
    className="u-mt-2 u-mb-half"
  >
    Add a new Todo
  </h2>
  <form
    onSubmit={[Function]}
  >
    <Label
      block={true}
      className=""
      error={false}
      htmlFor="todo-add-input"
    >
      Todo name
    </Label>
    <Input
      error={false}
      fullwidth={false}
      id="todo-add-input"
      onChange={[Function]}
      size="large"
      type="text"
      value=""
    />
    <DefaultButton
      busy={true}
      extension="narrow"
      label="add"
      size="large"
      type="submit"
    />
  </form>
</div>
`;
